File: match_rpc.go
// Copyright 2020 The Nakama Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"bytes"
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"time"

	"github.com/golang/protobuf/jsonpb"
	"github.com/heroiclabs/nakama-common/runtime"
	"github.com/heroiclabs/nakama-project-template/api"
)

func rpcFindMatch(marshaler *jsonpb.Marshaler, unmarshaler *jsonpb.Unmarshaler) func(context.Context, runtime.Logger, *sql.DB, runtime.NakamaModule, string) (string, error) {
	return func(ctx context.Context, logger runtime.Logger, db *sql.DB, nk runtime.NakamaModule, payload string) (string, error) {

		userID, ok := ctx.Value(runtime.RUNTIME_CTX_USER_ID).(string)
		if !ok {
			return "", errNoUserIdFound
		}

		exp, ok := ctx.Value(runtime.RUNTIME_CTX_USER_SESSION_EXP).(int64)
		if !ok || time.Now().Sub(time.Unix(exp, 0)) < 6*time.Hour {
			// 0 uses system expiry settings.
			exp = 0
		}

		vars, ok := ctx.Value(runtime.RUNTIME_CTX_VARS).(map[string]string)
		if !ok {
			vars = map[string]string{} // No session vars so set default.
		}

		users, err := nk.UsersGetId(ctx, []string{userID})
		if err != nil {
			logger.Error("UsersGetId error: %v", err)
			return "", errInternalError
		}

		// Use the latest username in the new token.
		token, exp, err := nk.AuthenticateTokenGenerate(userID, users[0].GetUsername(), exp, vars)
		if err != nil {
			logger.Error("AuthenticateTokenGenerate error: %v", err)
			return "", errInternalError
		}

		logger.Debug("New session with %d expiry time: %v", exp, token)

		var resp struct {
			Session  string   `json:"token"`
			MatchIds []string `protobuf:"bytes,1,rep,name=match_ids,json=matchIds,proto3" json:"match_ids,omitempty"`
		}
		resp.Session = token

		request := &api.RpcFindMatchRequest{}
		if err := unmarshaler.Unmarshal(bytes.NewReader([]byte(payload)), request); err != nil {
			return "", errUnmarshal
		}

		maxSize := 1
		var fast int
		if request.Fast {
			fast = 1
		}
		query := fmt.Sprintf("+label.open:1 +label.fast:%d", fast)

		matchIDs := make([]string, 0, 10)
		matches, err := nk.MatchList(ctx, 10, true, "", nil, &maxSize, query)
		if err != nil {
			logger.Error("error listing matches: %v", err)
			return "", errInternalError
		}
		if len(matches) > 0 {
			// There are one or more ongoing matches the user could join.
			for _, match := range matches {
				matchIDs = append(matchIDs, match.MatchId)
			}
		} else {
			// No available matches found, create a new one.
			matchID, err := nk.MatchCreate(ctx, moduleName, map[string]interface{}{"fast": request.Fast})
			if err != nil {
				logger.Error("error creating match: %v", err)
				return "", errInternalError
			}
			matchIDs = append(matchIDs, matchID)
		}

		resp.MatchIds = matchIDs

		out, err := json.Marshal(resp)
		if err != nil {
			logger.Error("Marshal error: %v", err)
			return "", errMarshal
		}

		logger.Debug("rpcFindMatch resp: %v", string(out))
		return string(out), nil
	}
}

// TODO: implement rpc function that returns the game state, precences and current tick, like the console does
// func (s *ConsoleServer) GetMatchState(ctx context.Context, in *console.MatchStateRequest) (*console.MatchState, error) {
// 	// Validate the match ID.
// 	matchIDComponents := strings.SplitN(in.GetId(), ".", 2)
// 	if len(matchIDComponents) != 2 {
// 		return nil, status.Error(codes.InvalidArgument, "Invalid match ID.")
// 	}
// 	matchID, err := uuid.FromString(matchIDComponents[0])
// 	if err != nil {
// 		return nil, status.Error(codes.InvalidArgument, "Invalid match ID.")
// 	}
// 	node := matchIDComponents[1]
// 	if node == "" {
// 		// Relayed matches don't have a state.
// 		return &console.MatchState{State: ""}, nil
// 	}

// 	presences, tick, state, err := s.matchRegistry.GetState(ctx, matchID, node)
// 	if err != nil {
// 		if err != context.Canceled && err != ErrMatchNotFound {
// 			s.logger.Error("Error getting match state.", zap.Any("in", in), zap.Error(err))
// 		}
// 		if err == ErrMatchNotFound {
// 			return nil, status.Error(codes.InvalidArgument, "Match not found, or match handler already stopped.")
// 		}
// 		return nil, status.Error(codes.Internal, "Error listing matches.")
// 	}

// 	return &console.MatchState{Presences: presences, Tick: tick, State: state}, nil
// }

func rpcGetMatch(marshaler *jsonpb.Marshaler, unmarshaler *jsonpb.Unmarshaler) func(context.Context, runtime.Logger, *sql.DB, runtime.NakamaModule, string) (string, error) {
	return func(ctx context.Context, logger runtime.Logger, db *sql.DB, nk runtime.NakamaModule, payload string) (string, error) {

		userID, ok := ctx.Value(runtime.RUNTIME_CTX_USER_ID).(string)
		if !ok {
			return "", errNoUserIdFound
		}

		// if len(payload) > 0 {
		// 	return "", errNoInputAllowed
		// }

		logger.Debug(payload)

		data, err := json.Marshal(payload)
		if err != nil {
			logger.Error("Marshal error: %v", err)
			return "", errMarshal
		}

		logger.Debug("payload data", data)
		// logger.Debug("payload data", json.Unmarshal(data))
		// logger.Debug(payload.matchId)

		// TODO: how  extend &api
		// request := &api.RpcGetMatchRequest{}
		// if err := unmarshaler.Unmarshal(bytes.NewReader([]byte(payload)), request); err != nil {
		// 	return "", errUnmarshal
		// }
		// logger.Debug(request)

		exp, ok := ctx.Value(runtime.RUNTIME_CTX_USER_SESSION_EXP).(int64)
		if !ok || time.Now().Sub(time.Unix(exp, 0)) < 6*time.Hour {
			// 0 uses system expiry settings.
			exp = 0
		}

		vars, ok := ctx.Value(runtime.RUNTIME_CTX_VARS).(map[string]string)
		if !ok {
			vars = map[string]string{} // No session vars so set default.
		}

		users, err := nk.UsersGetId(ctx, []string{userID})
		if err != nil {
			logger.Error("UsersGetId error: %v", err)
			return "", errInternalError
		}

		// Use the latest username in the new token.
		token, exp, err := nk.AuthenticateTokenGenerate(userID, users[0].GetUsername(), exp, vars)
		if err != nil {
			logger.Error("AuthenticateTokenGenerate error: %v", err)
			return "", errInternalError
		}

		logger.Debug("New session with %d expiry time: %v", exp, token)

		var resp struct {
			Session  string   `json:"token"`
			MatchIds []string `protobuf:"bytes,1,rep,name=match_ids,json=matchIds,proto3" json:"match_ids,omitempty"`
		}
		// resp.Session = token
		// resp.MatchIds = matchIDs

		out, err := json.Marshal(resp)
		if err != nil {
			logger.Error("Marshal error: %v", err)
			return "", errMarshal
		}

		return string(out), nil
		// return &MatchState{Presences: presences, Tick: tick, State: state}, nil
	}
}

File: match_rpc.ts
// Copyright 2020 The Nakama Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

let rpcFindMatch: nkruntime.RpcFunction = function (ctx: nkruntime.Context, logger: nkruntime.Logger, nk: nkruntime.Nakama, payload: string): string {
    if (!ctx.userId) {
        throw Error('No user ID in context');
    }

    if (!payload) {
        throw Error('Expects payload.');
    }

    let request = {} as RpcFindMatchRequest;
    try {
        request = JSON.parse(payload);
    } catch (error) {
        logger.error('Error parsing json message: %q', error);
        throw error;
    }

    let matches: nkruntime.Match[];
    try {
        const query = `+label.open:1 +label.fast:${request.fast ? 1 : 0}`;
        matches = nk.matchList(10, true, null, null, 1, query);
    } catch (error) {
        logger.error('Error listing matches: %v', error);
        throw error;
    }

    let match_ids: string[] = [];
    if (matches.length > 0) {
        // There are one or more ongoing matches the user could join.
        match_ids = matches.map(m => m.matchId);
    } else {
        // No available matches found, create a new one.
        try {
            match_ids.push(nk.matchCreate(moduleName, { fast: request.fast }));
        } catch (error) {
            logger.error('Error creating match: %v', error);
            throw error;
        }
    }

    // try {
    //     const user =  nk.usersGetId([ctx.userId])[0];
    // } catch (error) {
    //     logger.error('Error listing matches: %v', error);
    //     throw error;
    // }

    let res: RpcFindMatchResponse = { match_ids };
    return JSON.stringify(res);
}

let rpcListMatches: nkruntime.RpcFunction = function (ctx: nkruntime.Context, logger: nkruntime.Logger, nk: nkruntime.Nakama, payload: string): string {
    if (!ctx.userId) {
        throw Error('No user ID in context');
    }

    logger.info('ctx.userId', ctx.userId);

    // if (!payload) {
    //     throw Error('Expects payload.');
    // }

    let matches: nkruntime.Match[];
    try {
        const query = "+label.open:1";
        matches = nk.matchList(10, true, null, null, 1, query);
        logger.info('Matches', matches);
    } catch (error) {
        logger.error('Error listing matches: %v', error);
        throw error;
    }

    let match_ids: string[] = [];
    if (matches.length > 0) {
        // There are one or more ongoing matches the user could join.
        match_ids = matches.map(m => m.matchId);
    }

    let res: RpcFindMatchResponse = { match_ids };
    return JSON.stringify(res);
}

let rpcGeMatch: nkruntime.RpcFunction = function (ctx: nkruntime.Context, logger: nkruntime.Logger, nk: nkruntime.Nakama, payload: string): string {
    if (!ctx.userId) {
        throw Error('No user ID in context');
    }

    logger.info('rpcGeMatch() ctx.userId', ctx.userId);

    if (!payload) {
        throw Error('Expects payload.');
    }

    let state: State | null = null
    let match: Match | null = null;

    try {
        // logger.info('payload', payload);
        const data = JSON.parse(payload);
        match = nk.matchGet(data.matchId);
        state = {
            label: JSON.parse(match.label ? match.label : ''),
            emptyTicks: 0,
            presences: {},
            joinsInProgress: 0,
            playing: false,
            board: [],
            marks: {},
            mark: Mark.UNDEFINED,
            deadlineRemainingTicks: 0,
            winner: null,
            winnerPositions: null,
            nextGameRemainingTicks: 0,
        }

        // &{random:0xc001f9fef0 label:0xc000123b60 emptyTicks:0 presences:map[5bae80d3-6f42-44fb-9543-f3e1dc15b32d:0xc000c0e000] joinsInProgress:0 playing:false board:[] marks:map[] mark:0 deadlineRemainingTicks:0 winner:0 nextGameRemainingTicks:0}
        
    } catch (error) {
        logger.error('Error listing matches: %v', error);
        throw error;
    }

    let res: RpcGetMatchResponse = { match, state };
    return JSON.stringify(res);
}
